crates/bevy_diagnostic/src/entity_count_diagnostics_plugin.rs

```diff
@@ -0,0 +1,31 @@
+use crate::{Diagnostic, DiagnosticId, Diagnostics};
+use bevy_app::prelude::*;
+use bevy_ecs::{Entity, IntoSystem, Query, ResMut};
+
+/// Adds "entity count" diagnostic to an App
+#[derive(Default)]
+pub struct EntityCountDiagnosticsPlugin;
+
+impl Plugin for EntityCountDiagnosticsPlugin {
+    fn build(&self, app: &mut AppBuilder) {
+        app.add_startup_system(Self::setup_system.system())
+            .add_system(Self::diagnostic_system.system());
+    }
+}
+
+impl EntityCountDiagnosticsPlugin {
+    pub const ENTITY_COUNT: DiagnosticId =
+        DiagnosticId::from_u128(187513512115068938494459732780662867798);
+
+    pub fn setup_system(mut diagnostics: ResMut<Diagnostics>) {
+        diagnostics.add(Diagnostic::new(Self::ENTITY_COUNT, "entity_count", 20));
+    }
+
+    pub fn diagnostic_system(mut diagnostics: ResMut<Diagnostics>, query: Query<Entity>) {
+        let mut count = 0.;
+        for _entity in &mut query.iter() {
```

Comment:
we could replace this with `let count = query.iter().count()`. Thats simpler, but also if/when we re-implement ExactSizeIterator for query it could make this O(1)

Reply:
Or alternatively we could make it O(1) now by adding a world.entity_count() function and making this system thread-local.

Your call!

Reply:
Tried adding it to `World`, in my test with a system adding and removing entities every frame, the two counts are the same